### PR TITLE
Add high-frequency paper scalper engine and metrics endpoint

### PR DESCRIPTION
**Pull request by petradeweese**
## Summary
- add a shared scalper utility module for fees, sizing, metrics, and backtest summaries
- implement the high-frequency scalper engine with risk guards, exports, and backtest support while refactoring the LF engine to reuse shared helpers
- extend paper/settings UIs and routes with HF endpoints, shared metrics, and dual-mode charts/tables plus new HF tests

## Testing
- `pytest -q`

------
https://chatgpt.com/codex/tasks/task_e_68da905d0dd08329b697d7e76f9c27e2

### CODE REVIEW

**Comment by petradeweese** [comment]
Nice—this looks close. Before you merge the Scalper PR, drop this concise review request as a single GitHub comment. It asks Codex to surface all issues (not just the first) and to propose patches + tests.

⸻

PR comment for Codex

@codex Please run a full review on this PR and report all problems (don’t stop after the first). For each issue, propose a concrete fix (or apply it) plus tests. Validate the items below:

Engine & Risk
	•	Signal gating math and look-ahead bias checks for HF engine.
	•	Position sizing uses % of equity with per-trade cap and daily trade cap; caps enforced across restarts.
	•	Liquidity/slippage model (spread-aware) applied on both entry/exit; fees included in P&L.
	•	Hard/soft stops respected; max holding time enforced; no multiple fills per signal.

Lifecycle
	•	Start/Stop/Restart idempotent; status line updates with fresh timestamps after restart; state cleared on restart.
	•	Scheduler and dedupe prevent duplicate runs; rate-limit/backoff on data calls.

Data
	•	Schwab-first quotes/options via adapters; fallback only on 429/5xx or when disabled.
	•	Timestamps are exchange time and converted to local in UI consistently.

UI (Paper ▸ HF tab)
	•	Chart shows correct time ticks per range; ROI chip green/red as appropriate.
	•	Activity table columns: Ticker, Call/Put, Strike-Expiry (e.g., $555–9/14), Qty, Entry Time/Price, Exit Time/Price, Fees, ROI%.
	•	CSV export mirrors table schema; pagination/sorting stable.

Settings
	•	New HF settings persist: daily trade cap (default 30, editable), % of equity, stop/target, max hold.
	•	Validation + helpful errors; sensible defaults.

Testing & Quality
	•	Add/confirm: unit tests for signal gating, sizing, slippage/fees, caps, restart idempotency, CSV schema snapshot, settings persistence.
	•	End-to-end test that simulates a trading day with capped trades and verifies equity curve and logs.
	•	Run pytest -q (full pass), type check (mypy/pyright if configured), and lint (ruff/flake8).

If everything passes, proceed to merge; otherwise post a checklist of fixes with patches and updated tests.

**Comment by petradeweese** [comment]
@codex review and find bugs

**Comment by petradeweese** [comment]
@codex review

**Comment by petradeweese** [comment]
@codex review

File: services/scalper/lf_engine.py [hunk 1]
```diff
@@ -0,0 +1,889 @@
...
+    settings = load_settings(db)
+    rows = db.execute(
+        """
+        SELECT entry_time, exit_time, roi_pct, realized_pl, net_pl
+          FROM scalper_lf_activity
+         WHERE status='closed'
+         ORDER BY exit_time ASC
+        """
+    ).fetchall()
+    return dict(compute_trade_metrics(rows, starting_balance=settings.starting_balance))
```
[the remark this answers is not in the document]
Comment (petradeweese) [reply services/scalper/lf_engine.py:380]: @codex
Good catch. Please fix the metrics handling so LF and HF both work without errors:
	1.	LF Metrics Endpoint
	•	/api/paper/scalper/metrics.json?mode=lf currently breaks because sqlite3.Row objects are passed directly into compute_trade_metrics, which expects dict-like objects with .get.
	•	Fix: Convert rows to dicts before passing ([dict(r) for r in rows]) or update compute_trade_metrics to accept sqlite3.Row.
	2.	HF Metrics Endpoint
	•	/api/paper/scalper/metrics.json?mode=hf has the same issue: raw sqlite3.Row objects are passed into compute_trade_metrics, causing AttributeError once trades close.
	•	Fix: Same as LF — ensure rows are converted to dicts or update compute_trade_metrics to accept Row.

✅ Acceptance Criteria:
	•	Both endpoints (mode=lf and mode=hf) return proper JSON metrics after trades close.
	•	No AttributeError when rows exist.
	•	Unit tests cover the fix for both LF and HF.

**Comment by petradeweese** [comment]
@codex review

File: services/scalper/hf_engine.py [hunk 2]
```diff
@@ -0,0 +1,877 @@
...
+
+def _drawdown_breached(db, now_ts: datetime, settings: HFSettings) -> bool:
+    if settings.daily_max_drawdown_pct >= 0:
+        return False
+    current = _latest_equity(db)
+    session_start = _session_start_equity(db, now_ts.date(), default=settings.starting_balance)
+    if session_start <= 0:
+        return False
+    delta_pct = (current - session_start) / session_start * 100.0
+    return delta_pct <= settings.daily_max_drawdown_pct
```
[the remark this answers is not in the document]
Comment (petradeweese) [reply services/scalper/hf_engine.py:462]: @codex Good catch. Please update the daily-drawdown baseline logic:

Change
	•	_session_start_equity must use the most recent equity snapshot before the new session starts (or at restart_engine), not settings.starting_balance.
	•	On engine (re)start, persist a “session_baseline” row so all intraday checks reference that value.

Edge cases
	•	Fresh DB (no prior equity) → fall back to starting_balance.
	•	If the first trade of the day occurs before any intraday equity write, the baseline still equals the persisted session_baseline.

Tests
	•	Prior day ends at 80k on a 100k start with –6% daily cap; first trade next day does not halt (baseline = 80k).
	•	Fresh DB uses 100k start; baseline recorded at restart and read by guard.

**Comment by petradeweese** [comment]
@codex Thanks — please implement the following and push:

1) Session drawdown baseline
   • Change _session_start_equity to fall back to the most recent equity snapshot prior to session start (not settings.starting_balance).
   • On start/restart, persist the seeded baseline so intraday checks use that value consistently.
   • Update _drawdown_breached() to use the persisted baseline.
   • Tests: cover first-trade-of-day, post-restart, and prior-day loss/gain cases.

2) HF equity primary key collisions
   • Migrate HF equity ledger to allow multiple rows per timestamp using a composite PK (ts, trade_id). Update inserts to include trade_id.
   • Add a matching migration and backfill strategy; ensure LF path mirrors the approach if needed.
   • Tests: close ≥2 trades with same exit_time and assert both equity rows persist; no UNIQUE errors.

Finally, run full `pytest -q` and confirm UI/APIs (/paper/scalper metrics & status) still read from the new baseline and composite key without regressions.
If all tests pass and migrations apply cleanly, please merge.

**Comment by petradeweese** [comment]
@codex Please review and patch the following consolidated issues found in the scalper engines:
	1.	Session Baseline Handling (HF & LF)
	•	_session_start_equity falls back to settings.starting_balance when no equity row exists for the current day.
	•	This causes the daily drawdown guard to either trip too early (if the account grew) or fail to halt after losses (if the account shrank).
	•	Action: Persist the session baseline from the most recent equity snapshot (prior to session start/restart) instead of defaulting to the static seed.
	2.	Equity Table Primary Keys
	•	Current schema still uses ts TEXT PRIMARY KEY.
	•	This raises sqlite3.IntegrityError or silently overwrites when multiple closes occur in the same second.
	•	Action: Promote (ts, trade_id) to a composite key and add a migration to support backfills.
	3.	Metrics Endpoint Conversion
	•	Both HF and LF metrics_snapshot forward sqlite3.Row objects into compute_trade_metrics, which expects .get.
	•	This raises AttributeError: 'sqlite3.Row' object has no attribute 'get'.
	•	Action: Convert rows to dicts before passing them into compute_trade_metrics.
	4.	LF Metrics Snapshot
	•	Same issue as HF: sqlite3.Row passed without conversion, breaking /api/paper/scalper/metrics.json?mode=lf.
	•	Action: Cast to dict before computing metrics.
	5.	LF Status – Missing mark_price Column (P0)
	•	_unrealized_pl queries COALESCE(mark_price, entry_price), but scalper_lf_activity schema is missing mark_price.
	•	This throws sqlite3.OperationalError: no such column: mark_price.
	•	Action: Add mark_price via Alembic migration and backfill existing rows.

⸻

✅ Please apply fixes, add corresponding migrations, and re-run regression tests for HF/LF scalpers.
